tests/core/test_hookenv.py

```diff
@@ -1370,6 +1370,32 @@ def test_goal_state(self, check_output_):
         self.assertEqual(result, expect)
         check_output_.assert_called_with(['goal-state', '--format=json'])
 
+    @patch('subprocess.check_output')
+    def test_goal_state_filter_self_relations(self, check_output_):
+        data = {
+            'units': {'keystone/0': {}, 'keystone/1': {}},
+            'relations': {
+                'identity-service': {
+                    'glance/0': {},
+                    'keystone/0': {},
+                    'keystone/1': {},
+                },
+            },
+        }
+        expect = {
+            'units': {'keystone/0': {}, 'keystone/1': {}},
+            'relations': {
+                'identity-service': {
+                    'glance/0': {},
+                },
+            },
+        }
+        check_output_.return_value = json.dumps(data).encode('UTF-8')
```

Comment:
Personally, I think it would have been better to mock charmhelpers.core.hookenv.goal_state instead of subprocess.check_output. The way you have it, the tests are more fragile (eg. someone updating goal_state use different or tweaked subprocess calls would break these tests), and it is more complicated (goal_state.return_value = data, vs. serializing your data structure).

Reply:
Another approach would be to use functional composition instead and have `filter_self_relations` take the results of `goal_state` as input.  This would actually make the function name more accurate, but would make calling it even more verbose (though, again more explicit and self-documenting).